packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -117,12 +141,150 @@ export class UPlotChart extends React.Component<PlotProps, UPlotChartState> {
+    }
+
+    this.keyboardState.pressedKeys[e.key as SupportedKey] = false;
+
+    if (e.key === ' ') {
+      e.preventDefault();
+      e.stopPropagation();
+
+      // We do this so setSelect hooks get fired, zooming the plot
+      this.state.plot?.setSelect(this.state.plot.select);
+      this.keyboardState.dragStartX = null;
+    }
+  };
+
+  handleKeys = (e: React.KeyboardEvent) => {
+    if (!SUPPORTED_KEYS.includes(e.key as SupportedKey)) {
+      return;
+    }
+
+    e.preventDefault();
+    e.stopPropagation();
+
+    const newKey = !this.keyboardState.pressedKeys[e.key as SupportedKey];
+    if (newKey) {
+      const initiateAnimationLoop = Object.values(this.keyboardState.pressedKeys).every((pressed) => pressed === false);
+      this.keyboardState.pressedKeys[e.key as SupportedKey] = true;
+      this.keyboardState.dragStartX =
+        e.key === ' ' && this.keyboardState.dragStartX === null
+          ? this.state.plot?.cursor.left!
+          : this.keyboardState.dragStartX;
+
+      if (initiateAnimationLoop) {
+        window.requestAnimationFrame(this.handlePressedKeys);
+      }
+    }
+  };
+
+  handlePressedKeys = (time: number) => {
+    const nothingPressed = Object.values(this.keyboardState.pressedKeys).every((pressed) => pressed === false);
+    if (nothingPressed || !this.state.plot) {
+      this.keyboardState.keysLastHandledAt = null;
+      return;
+    }
+
+    const dt = time - (this.keyboardState.keysLastHandledAt ?? time);
+    const dx = dt * PIXELS_PER_MS;
+    let horValue = 0;
+    let vertValue = 0;
+
+    if (this.keyboardState.pressedKeys.ArrowUp) {
+      vertValue -= dx;
+    }
+    if (this.keyboardState.pressedKeys.ArrowDown) {
+      vertValue += dx;
+    }
+    if (this.keyboardState.pressedKeys.ArrowLeft) {
+      horValue -= dx;
+    }
+    if (this.keyboardState.pressedKeys.ArrowRight) {
+      horValue += dx;
+    }
+    if (this.keyboardState.pressedKeys.Shift) {
+      horValue *= SHIFT_MULTIPLIER;
+      vertValue *= SHIFT_MULTIPLIER;
+    }
+
+    this.moveCursor(horValue, vertValue);
+
+    const cursor = this.state.plot.cursor;
+    if (this.keyboardState.pressedKeys[' '] && cursor) {
+      const drawHeight = Number(this.state.plot.over.style.height.slice(0, -2));
+
+      this.state.plot.setSelect(
+        {
+          left: cursor.left! < this.keyboardState.dragStartX! ? cursor.left! : this.keyboardState.dragStartX!,
+          top: 0,
+          width: Math.abs(cursor.left! - (this.keyboardState.dragStartX ?? cursor.left!)),
+          height: drawHeight,
+        },
+        false
+      );
+    }
+
+    this.keyboardState.keysLastHandledAt = time;
+    window.requestAnimationFrame(this.handlePressedKeys);
+  };
+
+  moveCursor = (dx: number, dy: number) => {
+    if (this.state.plot?.cursor.left === undefined) {
+      return;
+    }
+
+    const { cursor } = this.state.plot;
+    this.state.plot.setCursor({
+      left: cursor.left! + dx,
+      top: cursor.top! + dy,
+    });
```

Comment:
Nit: can we constraint this to the plot canvas? One can move outside of the plot far far far away and got confused why things ain't working :)